### ISSUE TITLE
Set appropriate expiration for signed object URLs

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#3946]
When it generates a signed URL to download an artifact, the queue hard-codes that URL's expiration to 30 seconds.  The rational, AIUI, is that the caller should start the download immediately (and permission is checked at the start).

However, that means that if you open such a URL in a browser, the content displays but URL is stale 30 seconds later.

We could
* increase this value,
* make it configurable, or
* set it to the object's expiration

The advantage of the last option is that URLs in browsers will be ugly, but will not expire.  The justification is that we're allowing the caller to access the data, and if they want to disclose it to someone else then they can just download or copy/paste it.  So there's little sense in disallowing them to copy/paste the URL for that data.

@ricky26 what do you think?  @ajvb what is your reaction?

**Comment by ajvb** [comment]
Setting it to the object's expiration makes sense to me. What are object's expiration set to? Is that configurable?

**Comment by djmitche** [comment]
Yeah, it's configurable per-object.  Typically for task artifacts, the expiration is the same as the task's expiration, but can be earlier (for example we might keep binaries for the duration of the task but expire logs earlier, to save money)

**Comment by ajvb** [comment]
Ok totally. Is there a maximum length of time? What is the default?

>The justification is that we're allowing the caller to access the data, and if they want to disclose it to someone else then they can just download or copy/paste it.

Are there any cases where we'd want limits on this to protect against accidental disclosure? For example, do we have Mozilla Confidential artifacts that maybe should require SSO authentication or something similar?

**Comment by djmitche** [comment]
Yeah, that's the risk -- I think my argument is that "select-all / copy / slack / paste" isn't a much more unlikely path to accidental disclosure than "cmd-L / copy / slack paste".

**Comment by ricky26** [comment]
Using the object's expiry makes sense and sounds convenient to me. 👍 

I agree with @djmitche's comment that having a 'fully authorised' URL isn't actually much easier to expose compared to the file's content.

**Comment by djmitche** [comment]
We can definitely make this configurable per-backend.  But I'll make "object's lifetime" one of the options in that configuration.

**Comment by djmitche** [comment]
(thanks, both!)

**Comment by ajvb** [comment]
> Yeah, that's the risk -- I think my argument is that "select-all / copy / slack / paste" isn't a much more unlikely path to accidental disclosure than "cmd-L / copy / slack paste".

I'm not sure that I agree, but nevertheless I don't think that either protect against accidental exposure. But, having a URL being the primary mechanism for sharing does make protecting against accidental exposure easier.

How do you feel about opening a separate issue for discussing how to handle the case of protecting sensitive artifacts from accidental exposure? I don't think it's a blocker for this but I do feel like we should discuss and potentially address it.

**Comment by djmitche** [comment]
I don't think I understand what the concern is, so I'm not sure I can open an issue :)

**Comment by ajvb** [comment]
> I don't think I understand what the concern is, so I'm not sure I can open an issue :)

:)

My concern is that if we have artifacts that would be classified as Mozilla Confidential or stricter, that having them accessible from a public URL is dangerous particularly as it relates to accidental exposure of the URL. If an email get forwarded to a mailing list, or put in a public Matrix channel, etc etc, that could easily lead to a leak. My initial proposed solution would be to optionally support an authentication requirement on accessing these URLs.

As well, I am still curious if there is a maximum TTL for these artifacts and what the default TTL is?

**Comment by djmitche** [comment]
> My concern is that if we have artifacts that would be classified as Mozilla Confidential or stricter, that having them accessible from a public URL is dangerous particularly as it relates to accidental exposure of the URL. If an email get forwarded to a mailing list, or put in a public Matrix channel, etc etc, that could easily lead to a leak. My initial proposed solution would be to optionally support an authentication requirement on accessing these URLs.

These are download links, such as a URL for a built binary.  They are how users get the data *after* they have authenticated.  Adding another hoop of identification would be technically challenging (it'd need to be ergonomical to use with curl, and  supported by cloud backends like S3 or GCS, with no Taskcluster code involved) and only move the goalposts back a step: ok, now that you've authenticated _again_, you've got a URL for downloading the data (perhaps a signed S3 URL).  What if *that* URL is leaked?

The URLs in question are not exactly easy to leak -- they look like this:

https://community-tc.services.mozilla.com/api/queue/v1/task/BSIlVNx3QXmDQxH0GJDMfw/runs/0/artifacts/public%2Flogs%2Flive.log?bewit=Z2l0aHViLzI4NjczfGRqbWl0Y2hlXDE2MDU5MDA3MzdcMFJuSWw1QUEvLzAvQ2pXK01SSmlyVTVzNVd3aldKWmpJUjJ4ZDFxSXpDVT1cZXlKalpYSjBhV1pwWTJGMFpTSTZleUoyWlhKemFXOXVJam94TENKelkyOXdaWE1pT2xzaVlYTnpkVzFsT21kcGRHaDFZaTEwWldGdE9tMXZlbWxzYkdFdmFYUmZiM0JsY21GMGFXOXVjeUlzSW1GemMzVnRaVHBuYVhSb2RXSXRkR1ZoYlRwMFlYTnJZMngxYzNSbGNpOWpiM0psSWl3aVlYTnpkVzFsT21kcGRHaDFZaTEwWldGdE9tMXZlbWxzYkdFdlluVnBiR1F0Y21Wc1pXNW5ZWEJwTFhkeWFYUmxJaXdpWVhOemRXMWxPbWRwZEdoMVlpMTBaV0Z0T20xdmVtbHNiR0V2Ylc5NmFXeHNZUzFuWlc5c2IyTmhkR2x2YmlJc0ltRnpjM1Z0WlRwbmFYUm9kV0l0ZEdWaGJUcHRiM3BwYkd4aEwzUnlaV1ZvWlhKa1pYSWlMQ0poYzNOMWJXVTZaMmwwYUhWaUxYUmxZVzA2Ylc5NmFXeHNZUzloWkcxcGJpMWhiR3d0YjNKbkxXMWxiV0psY25NaUxDSmhjM04xYldVNloybDBhSFZpTFhSbFlXMDZiVzk2YVd4c1lTOTBZWE5yWTJ4MWMzUmxjaUlzSW1GemMzVnRaVHBuYVhSb2RXSXRkR1ZoYlRwMFlYTnJZMngxYzNSbGNpOXpaWEoyYVdObGN5MXlaWFpwWlhkbGNuTWlMQ0poYzNOMWJXVTZaMmwwYUhWaUxYUmxZVzA2ZEdGemEyTnNkWE4wWlhJdmQyOXlhMlZ5TFhKbGRtbGxkMlZ5Y3lJc0ltRnpjM1Z0WlRwbmFYUm9kV0l0ZEdWaGJUcDBZWE5yWTJ4MWMzUmxjaTkwWXkxaFpHMXBiaTF5WlhacFpYZGxjbk1pTENKaGMzTjFiV1U2WjJsMGFIVmlMWFJsWVcwNmJXOTZhV3hzWVM5amIyMXRkVzVwZEhrdGRHTXRZV1J0YVc1eklpd2lZWE56ZFcxbE9tZHBkR2gxWWkxMFpXRnRPblJoYzJ0amJIVnpkR1Z5TDNOdGIydGxkR1Z6ZEdWeWN5SXNJbUZ6YzNWdFpUcG5hWFJvZFdJdGRHVmhiVHAwWVhOclkyeDFjM1JsY2k5MFlYTnJZMngxYzNSbGNpMWliM1FpTENKaGMzTjFiV1U2WjJsMGFIVmlMWFJsWVcwNmJXOTZhV3hzWVM5elpYSjJhV05sTFdWdVoybHVaV1Z5YVc1bklpd2lZWE56ZFcxbE9tZHBkR2gxWWkxdmNtY3RZV1J0YVc0NmRHRnphMk5zZFhOMFpYSWlMQ0poYzNOMWJXVTZiRzluYVc0dGFXUmxiblJwZEhrNloybDBhSFZpTHpJNE5qY3pmR1JxYldsMFkyaGxJbDBzSW5OMFlYSjBJam94TmpBMU9EazRPRE16TlRNd0xDSmxlSEJwY25raU9qRTJNRFU1TURBMk16TTFNekFzSW5ObFpXUWlPaUptZEZOTlVWVlZNbE5wY1VGbFJWQmhUMnBHWkhGQlVHdFphRzUxYzAxUmNHVnJhakUyUjJOc1Z5MHhVU0lzSW5OcFoyNWhkSFZ5WlNJNklsTTBUa3hGUlRaak1qWmtUR1pvTUVwMlVIbHNkM1ZvVVhwdkwwZDVVa1ZWYkdaaWNFSnhXbU5FVjBFOUlpd2lhWE56ZFdWeUlqb2ljM1JoZEdsakwzUmhjMnRqYkhWemRHVnlMM2RsWWkxelpYSjJaWElpZlgwPQ

**Comment by ajvb** [comment]
> > My concern is that if we have artifacts that would be classified as Mozilla Confidential or stricter, that having them accessible from a public URL is dangerous particularly as it relates to accidental exposure of the URL. If an email get forwarded to a mailing list, or put in a public Matrix channel, etc etc, that could easily lead to a leak. My initial proposed solution would be to optionally support an authentication requirement on accessing these URLs.
> 
> These are download links, such as a URL for a built binary. They are how users get the data _after_ they have authenticated. Adding another hoop of identification would be technically challenging (it'd need to be ergonomical to use with curl, and supported by cloud backends like S3 or GCS, with no Taskcluster code involved) and only move the goalposts back a step: ok, now that you've authenticated _again_, you've got a URL for downloading the data (perhaps a signed S3 URL). What if _that_ URL is leaked?

I imagine there are a lot of artifacts that can absolutely be on publicly available URLs. What I am talking about is if Taskcluster is hosting artifacts that would be classified as Mozilla Confidential, then those should probably not be on publicly available URLs.

It would only move the goalpost in that way if that was how it was implemented. But to be honest, this is all getting into implementation details when I am still not sure if there is any risk at all or what the risk level is. Are there even artifacts that Taskcluster hosts that are Mozilla Confidential, Workgroup Confidential, etc? If so, how are they used?

At the same time, this discussion is not a blocker for this issue. So I'll go ahead and create a new issue surrounding this discussion.
